### PR TITLE
Added tests for IO subsystem

### PR DESCRIPTION
**Pull request by honza-kasik**

### CODE REVIEW

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import org.wildfly.extras.creaper.core.online.ModelNodeResult;
+import org.wildfly.extras.creaper.core.online.OnlineManagementClient;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.OperationException;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.admin.Administration;
+
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java:30]: I'm missing `org.junit.experimental.categories.Category` annotation specifying the EAP mode to run the tests against.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/WorkerTestCase.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+import org.wildfly.extras.creaper.core.online.operations.OperationException;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.admin.Administration;
+
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
+@RunWith(Arquillian.class)
+@RunAsClient
+public class WorkerTestCase {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/WorkerTestCase.java]: I'm missing `org.junit.experimental.categories.Category` annotation specifying the EAP mode to run the tests against.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            operations.removeIfExists(BUFFER_POOL_TBR_ADDRESS);
+        } finally {
+            administration.reloadIfRequired();
+        }
+    }
+
+    @Test
+    public void editBuffersPerSlice() throws Exception {
+        final int value = 10;
+        ModelNodeResult modelNodeResult = operations.readAttribute(BUFFER_POOL_ADDRESS, BUFFERS_PER_SLICE);
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java]: Please rename to something like `originalBuffersPerSliceResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                    .verifyAttribute(BUFFERS_PER_SLICE, value);
+        } finally {
+            operations.writeAttribute(BUFFER_POOL_ADDRESS, BUFFERS_PER_SLICE, modelNodeResult.value());
+        }
+    }
+
+    @Test
+    public void editBuffersPerSliceInvalid() throws Exception {
+        final int value = -1;
+        ModelNodeResult modelNodeResult = operations.readAttribute(BUFFER_POOL_ADDRESS, BUFFERS_PER_SLICE);
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java]: Please rename to something like `originalBuffersPerSliceResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                    .verifyFormNotSaved();
+        } finally {
+            operations.writeAttribute(BUFFER_POOL_ADDRESS, BUFFERS_PER_SLICE, modelNodeResult.value());
+        }
+    }
+
+    @Test
+    public void editBufferSize() throws Exception {
+        final int value = 1024;
+        ModelNodeResult modelNodeResult = operations.readAttribute(BUFFER_POOL_ADDRESS, BUFFER_SIZE);
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java]: Please rename to something like `originalBufferSizeResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                    .verifyAttribute(BUFFER_SIZE, value);
+        } finally {
+            operations.writeAttribute(BUFFER_POOL_ADDRESS, BUFFER_SIZE, modelNodeResult.value());
+        }
+    }
+
+    @Test
+    public void editBufferSizeInvalid() throws Exception {
+        final int value = -1;
+        ModelNodeResult originalModelNodeResult = operations.readAttribute(BUFFER_POOL_ADDRESS, BUFFER_SIZE);
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java]: Please rename to something like `originalBufferSizeResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        originalModelNodeResult.assertSuccess();
+        try {
+            new ConfigChecker.Builder(client, BUFFER_POOL_ADDRESS)
+                    .configFragment(page.getConfigFragment())
+                    .editAndSave(ConfigChecker.InputType.TEXT, BUFFER_SIZE, String.valueOf(value))
+                    .verifyFormNotSaved();
+        } finally {
+            operations.writeAttribute(BUFFER_POOL_ADDRESS, BUFFER_SIZE, originalModelNodeResult.value());
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java:139]: Please rename to something like `originalBufferSizeResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+    @Test
+    public void toggleDirectBuffers() throws Exception {
+        ModelNodeResult modelNodeResult = operations.readAttribute(BUFFER_POOL_ADDRESS, DIRECT_BUFFERS);
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/BufferPoolTestCase.java]: Please rename to something like `originalDirectBuffersResult` or at least `originalResult`.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/WorkerTestCase.java [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    private static final Administration administration = new Administration(client);
+    private static final Operations operations = new Operations(client);
+
+    private static final String
+            WORKER = "worker",
+            IO_THREADS = "io-threads",
+            TASK_KEEPALIVE = "task-keepalive",
+            TASK_MAX_THREADS = "task-max-threads",
+            STACK_SIZE = "stack-size",
+            FAIL_MESSAGE = "Probably fails because of https://issues.jboss.org/browse/HAL-1305. ";
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/io/WorkerTestCase.java]: Please rename to something more descriptive like `EDIT_FAIL_MESSAGE`

**Review by zebrik — CHANGES_REQUESTED** [review]
Please add the forgotten `@Category` annotations. The rest are just minor complaints.

**Comment by honza-kasik** [comment]
Thanks! I worked in requested changes!

**Comment by zebrik** [comment]
Thank you. Squash and I will merge.

**Comment by honza-kasik** [comment]
Squashed and ready to be merged. Thanks!

**Pull request merged**